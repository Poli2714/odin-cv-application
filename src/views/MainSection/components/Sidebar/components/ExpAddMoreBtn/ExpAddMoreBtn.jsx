import { useContext } from 'react';
import { v4 as uuidv4 } from 'uuid';

import { AddMoreBtn } from 'src/components/buttons';
import { isExpFormValid } from 'src/utils';
import {
  ExperienceContext,
  ExpErrMsgContext,
  KeyContext,
} from 'src/hooks/Contexts';

function ExpAddMoreBtn() {
  const { experiences, setExperiences } = useContext(ExperienceContext);
  const { setShowExpErrMessage } = useContext(ExpErrMsgContext);
  const { activeKey, setActiveKey } = useContext(KeyContext);
  const { expActiveKey } = activeKey;

  const handleClickAddExp = () => {
    const lastKey = Object.keys(experiences).at(-1);
    if (isExpFormValid(experiences[lastKey])) {
      const newExpID = uuidv4();
      setExperiences({
        ...experiences,
        [newExpID]: {
          jobTitle: '',
          employer: '',
          jobStartDate: '',
          jobEndDate: '',
          jobDescription: '',
        },
      });
      setActiveKey({ ...activeKey, expActiveKey: newExpID });
      setShowExpErrMessage(false);
    } else {
      expActiveKey !== lastKey &&
        setActiveKey({ ...activeKey, expActiveKey: lastKey });
      setShowExpErrMessage(true);
    }
  };

  return <AddMoreBtn onClick={handleClickAddExp} />;
}

export default ExpAddMoreBtn;
