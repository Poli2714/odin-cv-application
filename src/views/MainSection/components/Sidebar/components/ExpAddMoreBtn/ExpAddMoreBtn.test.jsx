import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { useState } from 'react';

import ExpAddMoreBtn from './ExpAddMoreBtn';
import {
  ExperienceContext,
  ExpErrMsgContext,
  KeyContext,
} from 'src/hooks/Contexts';

vi.mock('src/components/buttons', () => ({
  AddMoreBtn: vi.fn(({ onClick }) => (
    <button onClick={onClick}>Mock Add</button>
  )),
}));

function MockParentComponent() {
  const [experiences, setExperiences] = useState({ fakeID: { someData: '' } });
  const [activeKey, setActiveKey] = useState({ expActiveKey: '' });
  const [showExpErrMessage, setShowExpErrMessage] = useState(false);

  return (
    <ExperienceContext.Provider value={{ experiences, setExperiences }}>
      <KeyContext.Provider value={{ activeKey, setActiveKey }}>
        <ExpErrMsgContext.Provider
          value={{ showExpErrMessage, setShowExpErrMessage }}
        >
          <ExpAddMoreBtn />
        </ExpErrMsgContext.Provider>
      </KeyContext.Provider>
    </ExperienceContext.Provider>
  );
}

test('renders ExpAddMoreBtn', () => {
  render(<MockParentComponent />);
  expect(screen.getByRole('button')).toMatchSnapshot();
});
