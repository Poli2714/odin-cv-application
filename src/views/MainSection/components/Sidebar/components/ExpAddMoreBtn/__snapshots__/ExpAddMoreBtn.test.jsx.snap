// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders ExpAddMoreBtn 1`] = `
<button>
  Mock Add
</button>
`;
